File: backend/go/internal/metrics/metrics.go
// Package metrics is a minimal Prometheus text-format exporter. Metric names
// are identical in the Rust backend (docs/backend-deployment.md#metrics).
package metrics

import (
	"fmt"
	"io"
	"math"
	"sort"
	"sync"
	"sync/atomic"
)

type Counter struct{ v atomic.Uint64 }

func (c *Counter) Inc()         { c.v.Add(1) }
func (c *Counter) Add(n uint64) { c.v.Add(n) }
func (c *Counter) Get() uint64  { return c.v.Load() }

type Gauge struct{ v atomic.Int64 }

func (g *Gauge) Add(n int64) { g.v.Add(n) }
func (g *Gauge) Set(n int64) { g.v.Store(n) }
func (g *Gauge) Get() int64  { return g.v.Load() }

// FloatGauge stores a float64.
type FloatGauge struct{ v atomic.Uint64 }

func (g *FloatGauge) Set(f float64) { g.v.Store(math.Float64bits(f)) }
func (g *FloatGauge) Get() float64  { return math.Float64frombits(g.v.Load()) }

// LabeledCounter is a counter family with one label.
type LabeledCounter struct {
	mu sync.Mutex
	m  map[string]*Counter
}

func (l *LabeledCounter) With(label string) *Counter {
	l.mu.Lock()
	defer l.mu.Unlock()
	if l.m == nil {
		l.m = map[string]*Counter{}
	}
	c, ok := l.m[label]
	if !ok {
		c = &Counter{}
		l.m[label] = c
	}
	return c
}

func (l *LabeledCounter) snapshot() map[string]uint64 {
	l.mu.Lock()
	defer l.mu.Unlock()
	out := make(map[string]uint64, len(l.m))
	for k, c := range l.m {
		out[k] = c.Get()
	}
	return out
}

// Histogram with fixed buckets (seconds).
type Histogram struct {
	bounds []float64
	counts []atomic.Uint64
	sumNs  atomic.Uint64
	total  atomic.Uint64
}

func NewHistogram(bounds []float64) *Histogram {
	return &Histogram{bounds: bounds, counts: make([]atomic.Uint64, len(bounds))}
}

func (h *Histogram) Observe(seconds float64) {
	for i, b := range h.bounds {
		if seconds <= b {
			h.counts[i].Add(1)
		}
	}
	h.sumNs.Add(uint64(seconds * 1e9))
	h.total.Add(1)
}

// Registry holds all backend metrics.
type Registry struct {
	Implementation string
	Version        string

	ConnectedClients    Gauge
	ActiveVoiceSessions Gauge
	PacketsReceived     Counter
	PacketsSent         Counter
	BytesReceived       Counter
	BytesSent           Counter
	PacketsDropped      LabeledCounter
	InvalidPackets      LabeledCounter
	AuthFailures        Counter
	ControlMessages     Counter
	RelayLatency        *Histogram
	VoiceJitter         FloatGauge
}

func New(impl, version string) *Registry {
	return &Registry{Implementation: impl, Version: version,
		RelayLatency: NewHistogram([]float64{0.0001, 0.00025, 0.0005, 0.001, 0.0025, 0.005, 0.01, 0.025, 0.05})}
}

func (r *Registry) Write(w io.Writer) {
	p := func(format string, a ...any) { fmt.Fprintf(w, format, a...) }
	p("# HELP mcvoice_build_info Backend implementation and version.\n# TYPE mcvoice_build_info gauge\n")
	p("mcvoice_build_info{implementation=%q,version=%q} 1\n", r.Implementation, r.Version)
	gauge := func(name, help string, v int64) {
		p("# HELP %s %s\n# TYPE %s gauge\n%s %d\n", name, help, name, name, v)
	}
	counter := func(name, help string, v uint64) {
		p("# HELP %s %s\n# TYPE %s counter\n%s %d\n", name, help, name, name, v)
	}
	labeled := func(name, help, label string, l *LabeledCounter) {
		p("# HELP %s %s\n# TYPE %s counter\n", name, help, name)
		s := l.snapshot()
		keys := make([]string, 0, len(s))
		for k := range s {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		for _, k := range keys {
			p("%s{%s=%q} %d\n", name, label, k, s[k])
		}
	}
	gauge("mcvoice_connected_clients", "Control connections with an authenticated session.", r.ConnectedClients.Get())
	gauge("mcvoice_active_voice_sessions", "Sessions with a verified UDP voice path.", r.ActiveVoiceSessions.Get())
	counter("mcvoice_packets_received_total", "UDP datagrams received.", r.PacketsReceived.Get())
	counter("mcvoice_packets_sent_total", "UDP datagrams sent.", r.PacketsSent.Get())
	counter("mcvoice_bytes_received_total", "UDP bytes received.", r.BytesReceived.Get())
	counter("mcvoice_bytes_sent_total", "UDP bytes sent.", r.BytesSent.Get())
	labeled("mcvoice_packets_dropped_total", "Authenticated datagrams that were not relayed.", "reason", &r.PacketsDropped)
	labeled("mcvoice_invalid_packets_total", "Datagrams rejected as invalid.", "reason", &r.InvalidPackets)
	counter("mcvoice_auth_failures_total", "Failed control authentications.", r.AuthFailures.Get())
	counter("mcvoice_control_messages_total", "Control messages received.", r.ControlMessages.Get())
	p("# HELP mcvoice_voice_jitter_seconds Average interarrival jitter of active voice streams.\n# TYPE mcvoice_voice_jitter_seconds gauge\n")
	p("mcvoice_voice_jitter_seconds %g\n", r.VoiceJitter.Get())
	h := r.RelayLatency
	p("# HELP mcvoice_relay_latency_seconds Time from datagram receipt to relay send completion.\n# TYPE mcvoice_relay_latency_seconds histogram\n")
	for i, b := range h.bounds {
		p("mcvoice_relay_latency_seconds_bucket{le=\"%g\"} %d\n", b, h.counts[i].Load())
	}
	p("mcvoice_relay_latency_seconds_bucket{le=\"+Inf\"} %d\n", h.total.Load())
	p("mcvoice_relay_latency_seconds_sum %g\n", float64(h.sumNs.Load())/1e9)
	p("mcvoice_relay_latency_seconds_count %d\n", h.total.Load())
}

File: backend/rust/src/metrics.rs
//! Prometheus text exposition; metric names match backend/go.

use std::fmt::Write;
use std::sync::atomic::{AtomicI64, AtomicU64, Ordering::Relaxed};

pub const DROP_REASONS: &[&str] = &["no_hello", "not_routable", "rate_limited", "replay", "stale_epoch"];
pub const INVALID_REASONS: &[&str] = &[
    "auth_failed",
    "bad_counter",
    "bad_flags",
    "bad_magic",
    "bad_payload",
    "bad_version",
    "too_large",
    "too_short",
    "unknown_key",
    "unknown_session",
    "unknown_type",
    "uuid_mismatch",
];

const LATENCY_BOUNDS: [f64; 9] = [0.0001, 0.00025, 0.0005, 0.001, 0.0025, 0.005, 0.01, 0.025, 0.05];

#[derive(Default)]
pub struct Metrics {
    pub connected_clients: AtomicI64,
    pub active_voice_sessions: AtomicI64,
    pub packets_received: AtomicU64,
    pub packets_sent: AtomicU64,
    pub bytes_received: AtomicU64,
    pub bytes_sent: AtomicU64,
    pub auth_failures: AtomicU64,
    pub control_messages: AtomicU64,
    dropped: [AtomicU64; 5],
    invalid: [AtomicU64; 12],
    latency_buckets: [AtomicU64; 9],
    latency_sum_ns: AtomicU64,
    latency_count: AtomicU64,
    jitter_bits: AtomicU64,
}

impl Metrics {
    pub fn drop(&self, reason: &str) {
        if let Some(i) = DROP_REASONS.iter().position(|r| *r == reason) {
            self.dropped[i].fetch_add(1, Relaxed);
        }
    }

    pub fn invalid(&self, reason: &str) {
        if let Some(i) = INVALID_REASONS.iter().position(|r| *r == reason) {
            self.invalid[i].fetch_add(1, Relaxed);
        }
    }

    pub fn observe_latency(&self, secs: f64) {
        for (i, b) in LATENCY_BOUNDS.iter().enumerate() {
            if secs <= *b {
                self.latency_buckets[i].fetch_add(1, Relaxed);
            }
        }
        self.latency_sum_ns.fetch_add((secs * 1e9) as u64, Relaxed);
        self.latency_count.fetch_add(1, Relaxed);
    }

    pub fn set_jitter(&self, secs: f64) {
        self.jitter_bits.store(secs.to_bits(), Relaxed);
    }

    pub fn render(&self, version: &str) -> String {
        let mut o = String::with_capacity(4096);
        let _ = writeln!(
            o,
            "# HELP mcvoice_build_info Backend implementation and version.\n# TYPE mcvoice_build_info gauge"
        );
        let _ = writeln!(o, "mcvoice_build_info{{implementation=\"rust\",version=\"{version}\"}} 1");
        let gauge = |o: &mut String, n: &str, h: &str, v: i64| {
            let _ = writeln!(o, "# HELP {n} {h}\n# TYPE {n} gauge\n{n} {v}");
        };
        gauge(
            &mut o,
            "mcvoice_connected_clients",
            "Control connections with an authenticated session.",
            self.connected_clients.load(Relaxed),
        );
        gauge(
            &mut o,
            "mcvoice_active_voice_sessions",
            "Sessions with a verified UDP voice path.",
            self.active_voice_sessions.load(Relaxed),
        );
        let counter = |o: &mut String, n: &str, h: &str, v: u64| {
            let _ = writeln!(o, "# HELP {n} {h}\n# TYPE {n} counter\n{n} {v}");
        };
        counter(
            &mut o,
            "mcvoice_packets_received_total",
            "UDP datagrams received.",
            self.packets_received.load(Relaxed),
        );
        counter(
            &mut o,
            "mcvoice_packets_sent_total",
            "UDP datagrams sent.",
            self.packets_sent.load(Relaxed),
        );
        counter(
            &mut o,
            "mcvoice_bytes_received_total",
            "UDP bytes received.",
            self.bytes_received.load(Relaxed),
        );
        counter(&mut o, "mcvoice_bytes_sent_total", "UDP bytes sent.", self.bytes_sent.load(Relaxed));
        let _ = writeln!(o, "# HELP mcvoice_packets_dropped_total Authenticated datagrams that were not relayed.\n# TYPE mcvoice_packets_dropped_total counter");
        for (i, r) in DROP_REASONS.iter().enumerate() {
            let _ = writeln!(
                o,
                "mcvoice_packets_dropped_total{{reason=\"{r}\"}} {}",
                self.dropped[i].load(Relaxed)
            );
        }
        let _ = writeln!(
            o,
            "# HELP mcvoice_invalid_packets_total Datagrams rejected as invalid.\n# TYPE mcvoice_invalid_packets_total counter"
        );
        for (i, r) in INVALID_REASONS.iter().enumerate() {
            let _ = writeln!(
                o,
                "mcvoice_invalid_packets_total{{reason=\"{r}\"}} {}",
                self.invalid[i].load(Relaxed)
            );
        }
        counter(
            &mut o,
            "mcvoice_auth_failures_total",
            "Failed control authentications.",
            self.auth_failures.load(Relaxed),
        );
        counter(
            &mut o,
            "mcvoice_control_messages_total",
            "Control messages received.",
            self.control_messages.load(Relaxed),
        );
        let _ = writeln!(o, "# HELP mcvoice_voice_jitter_seconds Average interarrival jitter of active voice streams.\n# TYPE mcvoice_voice_jitter_seconds gauge");
        let _ = writeln!(o, "mcvoice_voice_jitter_seconds {}", f64::from_bits(self.jitter_bits.load(Relaxed)));
        let _ = writeln!(o, "# HELP mcvoice_relay_latency_seconds Time from datagram receipt to relay send completion.\n# TYPE mcvoice_relay_latency_seconds histogram");
        for (i, b) in LATENCY_BOUNDS.iter().enumerate() {
            let _ = writeln!(
                o,
                "mcvoice_relay_latency_seconds_bucket{{le=\"{b}\"}} {}",
                self.latency_buckets[i].load(Relaxed)
            );
        }
        let count = self.latency_count.load(Relaxed);
        let _ = writeln!(o, "mcvoice_relay_latency_seconds_bucket{{le=\"+Inf\"}} {count}");
        let _ = writeln!(
            o,
            "mcvoice_relay_latency_seconds_sum {}",
            self.latency_sum_ns.load(Relaxed) as f64 / 1e9
        );
        let _ = writeln!(o, "mcvoice_relay_latency_seconds_count {count}");
        o
    }
}
